src/include/index/skiplist.h

```diff
@@ -166,34 +185,198 @@ class SkipList {
       } else {
         curr_node = SearchFrom(key, curr_node, ctx).first->down_.load();
       }
+      // level_now is unsigned
+      if (level_now == 0) break;
       level_now--;
     }
   }
+
+  /*
+ * AddLevel() - add corresponding level to the SkipList
+ *
+ * return true if successfully added or the level is already added
+ * return false if the level cannot be reached from the highest level now
+ */
+  bool AddLevel(u_int32_t level) {
+    LOG_INFO("AddLevel");
+    SkipListBaseNode *head = this->skip_list_head_.load();
+    if (head->level_ + 1 < level) {
+      return false;
+    } else {
+      if (head->level_ + 1 == level) {
+        SkipListBaseNode *new_head = node_manager_.GetSkipListHead(level);
+        new_head->down_ = head;
+        if (this->skip_list_head_.compare_exchange_strong(head, new_head)) {
+          return true;
+        } else {
+          node_manager_.ReturnSkipListNode(new_head);
+          head = this->skip_list_head_.load();
+          return head->level_ == level;
+        }
+      } else {
+        return true;
+      }
+    }
+  }
+
+  /*
+   * InserNodeIntoInterval() - this method would try to insert the tower into
+   *the interval and retry due to contention
+   *
+   * It has a call_stack array to accelorate the process
+   * NOTE: this method would retry until the world ends (or the root is
+   *deleted)!!!
+   *
+   * Would only return true, or it would retry until succeed or the root is
+   *deleted
+   * Use this function only the tower can be exactly inserted
+   */
+  bool InsertTowerIntoInterval(
+      const KeyType &key, std::vector<SkipListInnerNode *> &tower,
+      std::vector<std::pair<SkipListBaseNode *, SkipListBaseNode *>> &
+          call_stack,
+      OperationContext &ctx, u_int32_t start_level = 0,
+      bool check_multiple_key_value=false) {
+    LOG_INFO("InsertTower");
+    u_int32_t expected_level = tower.size();
+    for (u_int32_t i = start_level; i < expected_level; i++) {
+      bool insert_flag = false;
+      do {
+        if (i != 0 && GET_DELETE(tower[i]->GetRoot().load()->next_.load())) {
+          // the root has been deleted
+          // there is no need to continue
+          for (auto j = i; j < expected_level; j++) {
+            node_manager_.ReturnSkipListNode(tower[j]);
+          }
+          return true;
+        }
+
+        //if the level is 0, multiple key-value pair is required
+        //if the check bool is true, then it should be non-unique index
+        if (i==0){
+          if (!check_multiple_key_value){
+            // unique index, just need to check the next one
+            if (!GET_DELETE(call_stack[i].second->next_.load())
+                &&ValueCmpEqual(tower[i]->GetValue(),
+                                static_cast<SkipListInnerNode *>(call_stack[i].second)->GetValue())){
+              return false;
+            }
+          }else{
+            //need to check all key-value pairs
```

Comment:
// There can be node with the same key and different values, we need to ensure no nodes with both key and value are the same